Improve quickstart5 (#2566)
PennyLaneAI/pennylane: doc/introduction/inspecting_circuits.rst

 .. role:: html(raw)
   :format: html

.. _intro_inspecting_circuits:

Inspecting circuits
===================

PennyLane offers functionality to inspect, visualize or analyze quantum circuits.

.. _intro_qtransforms:

> Most of these tools are implemented as **transforms**
> which take a :class:`~pennylane.QNode` instance and return a function:

Reviewer: ```suggestion
Most of these tools are implemented as **transforms**.  Transforms take a :class:`~pennylane.QNode` instance and return a function:
```

Not quite sure why I prefer this, but my gut is telling me it sounds a little better.

>>> @qml.qnode(dev, diff_method='parameter-shift')
... def my_qnode(x, a=True):
...     # ...
>>> new_func = my_transform(qnode)

This new function can
be called with the same arguments as the QNode and returns the desired outcome,
such as a dictionary of the QNode's properties, a matplotlib figure drawing the circuit,

> or a DAG representing its connectivity structure.

Reviewer 2: Really nice way of phrasing it 👍

'num_trainable_params': 4,
'num_device_wires': 4,
'device_name': 'default.qubit',
'expansion_strategy': 'gradient',
'gradient_options': {},
'interface': 'autograd',
'diff_method': 'parameter-shift',
'gradient_fn': 'pennylane.gradients.parameter_shift.param_shift',
'num_gradient_executions': 10}

Circuit drawing
---------------

PennyLane has two built-in circuit drawers, :func:`~pennylane.draw` and
:func:`~pennylane.draw_mpl`.

For example:

.. code-block:: python

    dev = qml.device('lightning.qubit', wires=(0,1,2,3))

    @qml.qnode(dev)
    def circuit(x, z):
        qml.QFT(wires=(0,1,2,3))
        qml.IsingXX(1.234, wires=(0,2))
        qml.Toffoli(wires=(0,1,2))
        qml.CSWAP(wires=(0,2,3))
        qml.RX(x, wires=0)
        qml.CRZ(z, wires=(3,0))
        return qml.expval(qml.PauliZ(0))

    fig, ax = qml.draw_mpl(circuit)(1.2345,1.2345)
    fig.show()

.. image:: ../_static/draw_mpl/main_example.png
    :align: center
    :width: 400px
    :target: javascript:void(0);

More information, including various fine-tuning options, can be found in
the :doc:`drawing module <../code/qml_drawer>`.

Debugging with mid-circuit snapshots
------------------------------------

> When debugging quantum circuits run on simulators, it is often useful to inspect the
> current quantum state in between gates.

Reviewer: What is "it"? Also quite passive.  Maybe try:

```suggestion
When debugging quantum circuits run on simulators, we may want to inspect the current quantum state between gates.
```

* `default.mixed`: each snapshot saves the density matrix
* `default.gaussian`: each snapshot saves the covariance matrix and vector of means

During normal execution, the snapshots are ignored:

.. code-block:: python

    dev = qml.device("default.qubit", wires=2)

    @qml.qnode(dev, interface=None)
    def circuit():
        qml.Snapshot()
        qml.Hadamard(wires=0)
        qml.Snapshot("very_important_state")
        qml.CNOT(wires=[0, 1])
        qml.Snapshot()
        return qml.expval(qml.PauliX(0))

However, when using the :func:`~pennylane.snapshots`
transform, intermediate device states will be stored and returned alongside the
results.

>>> qml.snapshots(circuit)()
{0: array([1.+0.j, 0.+0.j, 0.+0.j, 0.+0.j]),
'very_important_state': array([0.707+0.j, 0.+0.j, 0.707+0.j, 0.+0.j]),
2: array([0.707+0.j, 0.+0.j, 0.+0.j, 0.707+0.j]),
'execution_results': array(0.)}

DAG representation
------------------

The representation of a quantum circuit as a Directed Acyclic Graph (DAG) can be computed
using the :func:`~pennylane.commutation_dag` transform. In the DAG, each node represents
a quantum operation, and edges represent non-commutation

> between two operations.

Reviewer: I think it would be worth mentioning something along the lines of:
"This representation is used in the :func:`pattern_matching_optimization` and :func:`pattern_matching` transforms.  "

And end the section with:
```
For more details, see:

    * Iten, R., Moyard, R., Metger, T., Sutter, D., Woerner, S.
      "Exact and practical pattern matching for quantum circuit optimization" `doi.org/10.1145/3498325
      <https://dl.acm.org/doi/abs/10.1145/3498325>`_
```

Author: No strong opinion, but I would prefer not to link to where a transform is used, since this info may change?

Unless you want to make these two transforms visible too...do they need a separate section?

Reviewer: I just don't want to cause too much confusion between the `CommutationDAG` and the `CircuitGraph`, which is also a DAG. Just a different kind. There's more than one way to represent a circuit as a directed acyclic graph, and this isn't the obvious one, since it's more about commutation relationships.

Reviewer 2: Should we list both? There is also this:

https://pennylane.readthedocs.io/en/latest/code/api/pennylane.transforms.qcut.tape_to_graph.html

My main reason for suggesting we include graph representations of circuits here: we currently have a habit of re-doing this every time we need it. If we make these graph representations more visible, developers may recognize that one of the existing ones already fit their workflow.

Author: I beefed up the DAG section, will discuss with you on another channel
